docker/Dockerfile_full

```diff
@@ -1,64 +1,94 @@
-
-FROM python:3.6-slim
-
-SHELL ["/bin/bash", "-c"]
-
-RUN apt-get update -qq && \
-  apt-get install -y --no-install-recommends \
-  build-essential \
-  wget \
-  openssh-client \
-  graphviz-dev \
-  pkg-config \
-  git-core \
-  openssl \
-  libssl-dev \
-  libffi6 \
-  libffi-dev \
-  libpng-dev \
-  libpq-dev \
-  curl && \
-  apt-get clean && \
-  rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* && \
-  mkdir /install && \
-  mkdir /app
-
-WORKDIR /install
-
-# Copy as early as possible so we can cache ...
-COPY alt_requirements/ ./alt_requirements
-COPY requirements.txt .
 
-RUN pip install -r alt_requirements/requirements_full.txt
+# Create virtualenv to isolate builds
+RUN python -m venv /build
+
+# Install common libraries
+RUN apt-get update -qq \
+ && apt-get install -y --no-install-recommends \
+    # required by psycopg2 at build and runtime
+    libpq-dev \
+     # required for health check
+    curl \
+ && apt-get autoremove -y
+
+# Make sure we use the virtualenv
+ENV PATH="/build/bin:$PATH"
 
-COPY --from=builder /build/rasa.tar.gz .
-RUN pip install ./rasa.tar.gz[sql,spacy,mitie]
+# Stage to build and install everything
+FROM base as builder
 
+WORKDIR /src
+
+# Install all required build libraries
 RUN apt-get update -qq \
-    && apt-get install -y --no-install-recommends wget \
-    && wget -P /app/data/ https://s3-eu-west-1.amazonaws.com/mitie/total_word_feature_extractor.dat \
-    && apt-get remove -y wget \
-    && apt-get autoremove -y
+ && apt-get install -y --no-install-recommends \
+    build-essential \
+    wget \
+    openssh-client \
+    graphviz-dev \
+    pkg-config \
+    git-core \
+    openssl \
+    libssl-dev \
+    libffi6 \
+    libffi-dev \
+    libpng-dev
 
+# Install spacy models
 RUN pip install https://github.com/explosion/spacy-models/releases/download/en_core_web_md-2.1.0/en_core_web_md-2.1.0.tar.gz#egg=en_core_web_md==2.1.0 --no-cache-dir > /dev/null \
-    && python -m spacy link en_core_web_md en \
-    && pip install https://github.com/explosion/spacy-models/releases/download/de_core_news_sm-2.1.0/de_core_news_sm-2.1.0.tar.gz#egg=de_core_news_sm==2.1.0 --no-cache-dir > /dev/null \
-    && python -m spacy link de_core_news_sm de
+ && python -m spacy link en_core_web_md en \
+ && pip install https://github.com/explosion/spacy-models/releases/download/de_core_news_sm-2.1.0/de_core_news_sm-2.1.0.tar.gz#egg=de_core_news_sm==2.1.0 --no-cache-dir > /dev/null \
+ && python -m spacy link de_core_news_sm de
+
+# Copy only what we really need
+COPY README.md .
+COPY setup.py .
+COPY MANIFEST.in .
+COPY alt_requirements/ ./alt_requirements
+COPY requirements.txt .
+
+# Install Rasa and its dependencies
+RUN pip install --no-cache-dir -r alt_requirements/requirements_full.txt
 
+# Install Rasa as package
+COPY rasa ./rasa
+RUN pip install .[sql,spacy,mitie]
+
+# Runtime stage which uses the virtualenv which we built in the previous stage
+FROM base AS runner
+
+# Install mitie model
```

Comment:
can't the mitie installation happen in the `builder` stage?

Reply:
and then just
```suggestion
COPY --from=builder /src/total_word_feature_extractor.dat .
```

Reply:
good point! Will do that